infra/modules/bootstrap/main.tf

```diff
@@ -0,0 +1,252 @@
+data "aws_caller_identity" "current" {}
+
+data "aws_partition" "current" {}
+
+# Create the dynamodb table requried for state locking.
+resource "aws_kms_key" "terraform_lock" {
+  description             = "KMS key for DynamoDB"
+  deletion_window_in_days = "10"
+  enable_key_rotation     = "true"
+}
+
+resource "aws_dynamodb_table" "terraform_lock" {
+  name         = var.dynamodb_table
+  hash_key     = "LockID"
+  billing_mode = "PAY_PER_REQUEST"
+
+  attribute {
+    name = "LockID"
+    type = "S"
+  }
+
+  server_side_encryption {
+    enabled     = true
+    kms_key_arn = aws_kms_key.terraform_lock.arn
+  }
+
+  point_in_time_recovery {
+    enabled = true
+  }
+
+}
+
+# Create the S3 bucket used to store terraform state remotely.
+resource "aws_s3_bucket" "tf_state" {
+  bucket = var.state_bucket_name
+
+  # Prevent accidental destruction a developer executing terraform destory in the wrong directory. Contains terraform state files.
+  lifecycle {
+    prevent_destroy = true
+  }
+}
+
+resource "aws_s3_bucket_versioning" "tf_state" {
+  bucket = aws_s3_bucket.tf_state.id
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "tf_state" {
+  bucket = aws_s3_bucket.tf_state.id
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm = "AES256"
```

Comment:
Terraform can still contain sensitive information (such as the initial db user and password) when creating a resource. We could use KMS keys here, I chose to use S3 Bucket Keys because its free. I could add both KMS (to fine tune control) and an S3 Bucket Key (which can reduce request costs).

https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-key.html

Reply:
oh cool, thanks for the link! didn't know about this 👍